featurebyte/api/user_defined_function_injector.py

```diff
@@ -0,0 +1,331 @@
+    @staticmethod
+    def _extract_parameter_value(
+        pos: int, func_param: FunctionParameter, *args: Any, **kwargs: Any
+    ) -> Tuple[Any, Any]:
+        if pos < len(args):
+            value = args[pos]
+        else:
+            if func_param.name not in kwargs:
+                if func_param.has_default_value:
+                    value = func_param.default_value
+                else:
+                    raise ValueError(f'Parameter "{func_param.name}" is not provided.')
+            else:
+                value = kwargs.pop(func_param.name)
+        return value, kwargs
+
+    @staticmethod
+    def _validate_series_inputs(series_inputs: FuncInputSeriesList) -> None:
+        expected_row_index_lineage = series_inputs[0].row_index_lineage
+        check_row_index_lineage = not isinstance(series_inputs[0], Feature)
+        expected_series_type = Feature if isinstance(series_inputs[0], Feature) else ViewColumn
+        for series_input in series_inputs[1:]:
+            if not isinstance(series_input, expected_series_type):
+                # check if all view columns have matching type
+                raise TypeError(
+                    f'Input "{series_input.name}" has type {type(series_input).__name__} '
+                    f"but expected {expected_series_type.__name__}."
+                )
+
+            if (
+                check_row_index_lineage
+                and series_input.row_index_lineage != expected_row_index_lineage
+            ):
+                # check if all view columns have matching row index lineage
+                raise ValueError(
+                    f'The row of the input ViewColumns "{series_inputs[0].name}" does not match '
+                    f'the row of the input ViewColumns "{series_input.name}".'
+                )
+
+    def _extract_node_parameters(
+        self, *args: Any, **kwargs: Any
+    ) -> Tuple[List[FunctionParameterInput], FuncInputSeriesList]:
+        if len(args) > len(self.function_parameters):
+            raise ValueError(
+                f"Too many arguments. Expected {len(self.function_parameters)} but got {len(args)}."
+            )
+
+        # extract input parameters based on the function parameter specification
+        func_params = []
+        series_inputs = []
+        value: Any
+        for i, param in enumerate(self.function_parameters):
+            # extract the parameter value from the arguments
+            value, kwargs = self._extract_parameter_value(i, param, *args, **kwargs)
+            check_type(
+                param.name, value=value, expected_type=get_param_type_annotations(param.dtype)
+            )
+
+            # prepare generic function parameter
+            func_param: FunctionParameterInput
+            if isinstance(value, (ViewColumn, Feature)):
+                if value.dtype != param.dtype:
+                    raise TypeError(
+                        f'Input ViewColumn or Feature "{value.name}" has dtype {value.dtype} '
```

Comment:
Will the `value.name` be empty if the column or feature is unnamed? In that case might be useful to mention the name of the violated argument in the error message.